File: katas/java/SOLID/dip/DipTests.java
package SOLID.dip;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests para el principio de Inversión de Dependencias (DIP)
 * Valida que las clases de alto nivel no dependan de clases de bajo nivel, sino de abstracciones
 */
public class DipTests {
    
    @Test
    public void testUserService_WithMySQLDatabase() {
        // Arrange
        Database database = new MySQLDatabase();
        UserService userService = new UserService(database);
        
        // Act
        String result = userService.getUser(1);
        
        // Assert
        assertNotNull(result, "Debe poder obtener usuario desde MySQL");
        assertTrue(result.contains("MySQL"), "El resultado debe indicar que viene de MySQL");
    }
    
    @Test
    public void testUserService_WithMongoDatabase() {
        // Arrange
        Database database = new MongoDatabase();
        UserService userService = new UserService(database);
        
        // Act
        String result = userService.getUser(1);
        
        // Assert
        assertNotNull(result, "Debe poder obtener usuario desde MongoDB");
        assertTrue(result.contains("MongoDB"), "El resultado debe indicar que viene de MongoDB");
    }
    
    @Test
    public void testUserService_WithPostgreSQLDatabase() {
        // Arrange
        Database database = new PostgreSQLDatabase();
        UserService userService = new UserService(database);
        
        // Act
        String result = userService.getUser(1);
        
        // Assert
        assertNotNull(result, "Debe poder obtener usuario desde PostgreSQL");
        assertTrue(result.contains("PostgreSQL"), "El resultado debe indicar que viene de PostgreSQL");
    }
    
    @Test
    public void testDIP_UserServiceDependsOnAbstraction() {
        // Arrange - UserService depende de la interfaz Database, no de implementaciones concretas
        Database mysql = new MySQLDatabase();
        Database mongo = new MongoDatabase();
        Database postgres = new PostgreSQLDatabase();
        
        // Act - UserService puede trabajar con cualquier implementación de Database
        UserService service1 = new UserService(mysql);
        UserService service2 = new UserService(mongo);
        UserService service3 = new UserService(postgres);
        
        // Assert - Todas las implementaciones funcionan con UserService
        assertNotNull(service1.getUser(1), "MySQL debe funcionar");
        assertNotNull(service2.getUser(1), "MongoDB debe funcionar");
        assertNotNull(service3.getUser(1), "PostgreSQL debe funcionar");
    }
    
    @Test
    public void testDIP_DatabaseImplementationsFollowContract() {
        // Arrange
        Database mysql = new MySQLDatabase();
        Database mongo = new MongoDatabase();
        
        // Act
        String mysqlUser = mysql.getUser(1);
        String mongoUser = mongo.getUser(1);
        
        // Assert - Todas las implementaciones cumplen el contrato de Database
        assertNotNull(mysqlUser, "MySQL debe implementar getUser");
        assertNotNull(mongoUser, "MongoDB debe implementar getUser");
    }
    
    @Test
    public void testMySQLDatabase_CanBeInstantiated() {
        // Arrange & Act
        MySQLDatabase database = new MySQLDatabase();
        
        // Assert
        assertNotNull(database, "MySQLDatabase debe poder ser instanciada");
        assertTrue(database instanceof Database, "MySQLDatabase debe implementar Database");
    }
    
    @Test
    public void testMongoDatabase_CanBeInstantiated() {
        // Arrange & Act
        MongoDatabase database = new MongoDatabase();
        
        // Assert
        assertNotNull(database, "MongoDatabase debe poder ser instanciada");
        assertTrue(database instanceof Database, "MongoDatabase debe implementar Database");
    }
}

File: katas/php/SOLID/dip/DipTests.php
<?php

use PHPUnit\Framework\TestCase;

require_once 'dip-good.php';

/**
 * Tests para el principio de Inversión de Dependencias (DIP)
 * Valida que las clases de alto nivel no dependan de clases de bajo nivel, sino de abstracciones
 */
class DipTests extends TestCase
{
    public function testUserService_WithMySQLDatabase()
    {
        // Arrange
        $database = new MySQLDatabase();
        $userService = new UserService($database);
        
        // Act
        $result = $userService->getUser(1);
        
        // Assert
        $this->assertNotNull($result, 'Debe poder obtener usuario desde MySQL');
        $this->assertStringContainsString("MySQL", $result, 'El resultado debe indicar que viene de MySQL');
    }
    
    public function testUserService_WithMongoDatabase()
    {
        // Arrange
        $database = new MongoDatabase();
        $userService = new UserService($database);
        
        // Act
        $result = $userService->getUser(1);
        
        // Assert
        $this->assertNotNull($result, 'Debe poder obtener usuario desde MongoDB');
        $this->assertStringContainsString("MongoDB", $result, 'El resultado debe indicar que viene de MongoDB');
    }
    
    public function testUserService_WithPostgreSQLDatabase()
    {
        // Arrange
        $database = new PostgreSQLDatabase();
        $userService = new UserService($database);
        
        // Act
        $result = $userService->getUser(1);
        
        // Assert
        $this->assertNotNull($result, 'Debe poder obtener usuario desde PostgreSQL');
        $this->assertStringContainsString("PostgreSQL", $result, 'El resultado debe indicar que viene de PostgreSQL');
    }
    
    public function testDIP_UserServiceDependsOnAbstraction()
    {
        // Arrange - UserService depende de la interfaz Database, no de implementaciones concretas
        $mysql = new MySQLDatabase();
        $mongo = new MongoDatabase();
        $postgres = new PostgreSQLDatabase();
        
        // Act - UserService puede trabajar con cualquier implementación de Database
        $service1 = new UserService($mysql);
        $service2 = new UserService($mongo);
        $service3 = new UserService($postgres);
        
        // Assert - Todas las implementaciones funcionan con UserService
        $this->assertNotNull($service1->getUser(1), 'MySQL debe funcionar');
        $this->assertNotNull($service2->getUser(1), 'MongoDB debe funcionar');
        $this->assertNotNull($service3->getUser(1), 'PostgreSQL debe funcionar');
    }
    
    public function testDIP_DatabaseImplementationsFollowContract()
    {
        // Arrange
        $mysql = new MySQLDatabase();
        $mongo = new MongoDatabase();
        
        // Act
        $mysqlUser = $mysql->getUser(1);
        $mongoUser = $mongo->getUser(1);
        
        // Assert - Todas las implementaciones cumplen el contrato de Database
        $this->assertNotNull($mysqlUser, 'MySQL debe implementar getUser');
        $this->assertNotNull($mongoUser, 'MongoDB debe implementar getUser');
    }
    
    public function testMySQLDatabase_CanBeInstantiated()
    {
        // Arrange & Act
        $database = new MySQLDatabase();
        
        // Assert
        $this->assertNotNull($database, 'MySQLDatabase debe poder ser instanciada');
        $this->assertInstanceOf(Database::class, $database, 'MySQLDatabase debe implementar Database');
    }
    
    public function testMongoDatabase_CanBeInstantiated()
    {
        // Arrange & Act
        $database = new MongoDatabase();
        
        // Assert
        $this->assertNotNull($database, 'MongoDatabase debe poder ser instanciada');
        $this->assertInstanceOf(Database::class, $database, 'MongoDatabase debe implementar Database');
    }
}
